### PR TITLE
Bandwidth watch

### PR DESCRIPTION
**Pull request by brianolson**
## Summary

Expand test/heapwatch/heapWatch.py to also gather /metrics and block headers.
Add a reporting script to measure differences across metrics and blocks.

## Test Plan

```sh
rm -rf ~/Algorand/tn3/
bash test/heapwatch/bwstart.sh ~/Algorand/tn3/
sleep 950
bash test/heapwatch/stop.sh ~/Algorand/tn3/

python3 test/heapwatch/metrics_delta.py --report ~/Algorand/tn3/report.csv --deltas ~/Algorand/tn3/txsd.csv ~/Algorand/tn3/heaps/Primary*metrics
```

(run again with test/heapwatch/* scripts copied onto branch `feature/txnsync`, compare `algod_network_sent_bytes_*` )

### CODE REVIEW

**Comment by brianolson** [comment]
For comparable TPS, I'm seeng `master` Primary relay node sending about 440 kB/s and for branch `feature/txnsync` I'm seeing that node send at about 530 kB/s

**Comment by tsachiherman** [comment]
> For comparable TPS, I'm seeng `master` Primary relay node sending about 440 kB/s and for branch `feature/txnsync` I'm seeing that node send at about 530 kB/s

I assume that you've been using the `ThreeNodesEvenDist.json` network, right ? on this network, you have one relay and two non-relays. On this network, it's expected to have slightly higher outgoing bandwidth consumption due to the bloom filter sending overhead. This network also doesn't "suffer" from multiple messages, since the number of relays is exactly one. You will have duplicate messages once you start having more than a single relay.

File: gen/generate.go [hunk 1]
```diff
@@ -79,7 +90,26 @@ func GenerateGenesisFiles(genesisData GenesisData, consensus config.ConsensusPro
...
+		if (math.Abs((fsum-ftot)/ftot) < 0.01) && (u64absDiff(sum, TotalMoney) < 10000) {
+			// wallet stake is a float and roundoff might happen but we might be close enough to do fixup
+			i := 0
+			for sum != TotalMoney {
+				if sum < TotalMoney {
+					allocation[i].Stake += 1
+					sum += 1
+				} else {
+					if allocation[i].Stake > 1 {
+						allocation[i].Stake -= 1
```
Comment (tsachiherman) [inline gen/generate.go:104]: Make sure not to drop balance below minbalance.

File: gen/generate.go [hunk 2]
```diff
@@ -79,7 +90,26 @@ func GenerateGenesisFiles(genesisData GenesisData, consensus config.ConsensusPro
 	}
 
 	if sum != TotalMoney {
-		panic(fmt.Sprintf("Amounts don't add up to TotalMoney - off by %v", int64(TotalMoney)-int64(sum)))
+		fsum := float64(sum)
```
Comment (tsachiherman) [inline gen/generate.go:93]: I'm not sure if this is the best place to do this adjustment, but the idea is good.

Comment (brianolson) [reply gen/generate.go:114]: GenesisData.Wallets[].Stake is a float64 [0..100.0], this is where we convert it to uint64 microalgos, so I think it fits in this function.

Comment (tsachiherman) [reply gen/generate.go:114]: sorry - my bad. I missed the `Stake:  uint64(float64(TotalMoney/100)*wallet.Stake + .5)` above, and thought it was implemented elsewhere.

**Review by tsachiherman — COMMENTED** [review]
could you move the `gen/generate.go` changes into it's own PR, and add Shiqi to that review as well ?

**Comment by brianolson** [comment]
I'm happy with how this has been running in recent tests vs feature/txnsync; ready to merge this chunk.

**Review by tsachiherman — APPROVED** [review]
looks good.

**Pull request merged**